File: templates/Enterprise-Template/src/typescript/generator-botbuilder-enterprise/generators/app/templates/enterprise-bot/src/dialogs/main/mainResponses.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import {
    ActionTypes,
    Activity,
    Attachment,
    CardFactory,
    InputHints,
    MessageFactory,
    TurnContext } from 'botbuilder';
import * as i18n from 'i18n';
import {
    DictionaryRenderer,
    LanguageTemplateDictionary,
    TemplateFunction } from '../templateManager/dictionaryRenderer';
import { TemplateManager } from '../templateManager/templateManager';

export class MainResponses extends TemplateManager {

    // Fields
    public static RESPONSE_IDS: {
        Cancelled: string;
        Completed: string;
        Confused: string;
        Greeting: string;
        Help: string;
        Intro: string;
    } = {
        Cancelled: 'cancelled',
        Completed: 'completed',
        Confused: 'confused',
        Greeting:  'greeting',
        Help:  'help',
        Intro:  'intro'
    };
    private static readonly RESPONSE_TEMPLATES: LanguageTemplateDictionary = new Map([
        ['default', new Map([
            [MainResponses.RESPONSE_IDS.Cancelled, MainResponses.fromResources('main.cancelled')],
            [MainResponses.RESPONSE_IDS.Completed, MainResponses.fromResources('main.completed')],
            [MainResponses.RESPONSE_IDS.Confused, MainResponses.fromResources('main.confused')],
            [MainResponses.RESPONSE_IDS.Greeting, MainResponses.fromResources('main.greeting')],
            [MainResponses.RESPONSE_IDS.Help,
                // tslint:disable-next-line:no-any
                (context: TurnContext, data: any): Promise<Activity> => MainResponses.BUILD_HELP_CARD(context, data)],
            [MainResponses.RESPONSE_IDS.Intro,
                // tslint:disable-next-line:no-any
                (context: TurnContext, data: any): Promise<Activity> => MainResponses.BUILD_INTRO_CARD(context, data)]
        ])]
    ]);

    constructor() {
        super();
        this.register(new DictionaryRenderer(MainResponses.RESPONSE_TEMPLATES));
    }

    // tslint:disable-next-line:no-any
    public static BUILD_INTRO_CARD(turnContext: TurnContext, data: any): Promise<Activity> {
        const introPath: string = i18n.__('main.introPath');
        // tslint:disable-next-line:no-any non-literal-require
        const introCard: any = require(introPath);
        const attachment: Attachment = CardFactory.adaptiveCard(introCard);
        const response: Partial<Activity> = MessageFactory.attachment(attachment, '', attachment.content.speak, InputHints.AcceptingInput);

        response.suggestedActions = {
            actions: [
                {
                    title: i18n.__('main.helpBtnText1'),
                    type: ActionTypes.ImBack,
                    value: i18n.__('main.helpBtnValue1')
                },
                {
                    title: i18n.__('main.helpBtnText2'),
                    type: ActionTypes.ImBack,
                    value: i18n.__('main.helpBtnValue2')
                },
                {
                    title: i18n.__('main.helpBtnText3'),
                    type: ActionTypes.OpenUrl,
                    value: i18n.__('main.helpBtnValue3')
                }
            ],
            to: []
        };

        return Promise.resolve(<Activity> response);
    }

    // tslint:disable-next-line:no-any
    public static async BUILD_HELP_CARD(turnContext: TurnContext, data: any): Promise<Activity> {
        const title: string = i18n.__('main.helpTitle');
        const text: string = i18n.__('main.helpText');
        const attachment: Attachment = CardFactory.heroCard(title, text);
        const response: Partial<Activity> = MessageFactory.attachment(attachment, text, InputHints.AcceptingInput);

        response.suggestedActions = {
            actions: [
            {
                title: i18n.__('main.helpBtnText1'),
                type: ActionTypes.ImBack,
                value: i18n.__('main.helpBtnValue1')
            },
            {
                title: i18n.__('main.helpBtnText2'),
                type: ActionTypes.ImBack,
                value: i18n.__('main.helpBtnValue2')
            },
            {
                title: i18n.__('main.helpBtnText3'),
                type: ActionTypes.OpenUrl,
                value: i18n.__('main.helpBtnValue3')
            }
            ],
            to: []
        };

        return Promise.resolve(<Activity> response);
    }

    private static fromResources(name: string): TemplateFunction {
        return (): Promise<string> => Promise.resolve(i18n.__(name));
    }
}

File: templates/Enterprise-Template/src/csharp/EnterpriseBotSample/EnterpriseBotSample/Dialogs/Main/MainResponses.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.IO;
using AdaptiveCards;
using EnterpriseBotSample.Dialogs.Main.Resources;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.TemplateManager;
using Microsoft.Bot.Schema;

namespace EnterpriseBotSample.Dialogs.Main
{
    public class MainResponses : TemplateManager
    {
        private static LanguageTemplateDictionary _responseTemplates = new LanguageTemplateDictionary
        {
            ["default"] = new TemplateIdMap
            {
                { ResponseIds.Cancelled,
                    (context, data) =>
                    MessageFactory.Text(
                        text: MainStrings.CANCELLED,
                        ssml: MainStrings.CANCELLED,
                        inputHint: InputHints.AcceptingInput)
                },
                { ResponseIds.Completed,
                    (context, data) =>
                    MessageFactory.Text(
                        text: MainStrings.COMPLETED,
                        ssml: MainStrings.COMPLETED,
                        inputHint: InputHints.AcceptingInput)
                },
                { ResponseIds.Confused,
                    (context, data) =>
                    MessageFactory.Text(
                        text: MainStrings.CONFUSED,
                        ssml: MainStrings.CONFUSED,
                        inputHint: InputHints.AcceptingInput)
                },
                { ResponseIds.Greeting,
                    (context, data) =>
                    MessageFactory.Text(
                        text: MainStrings.GREETING,
                        ssml: MainStrings.GREETING,
                        inputHint: InputHints.AcceptingInput)
                },
                { ResponseIds.Help, (context, data) => BuildHelpCard(context, data) },
                { ResponseIds.Intro, (context, data) => BuildIntroCard(context, data) },
            }
        };

        public MainResponses()
        {
            Register(new DictionaryRenderer(_responseTemplates));
        }

        public static IMessageActivity BuildIntroCard(ITurnContext turnContext, dynamic data)
        {
            var introCard = File.ReadAllText(MainStrings.INTRO_PATH);
            var card = AdaptiveCard.FromJson(introCard).Card;
            var attachment = new Attachment(AdaptiveCard.ContentType, content: card);

            var response = MessageFactory.Attachment(attachment, ssml: card.Speak, inputHint: InputHints.AcceptingInput);

            response.SuggestedActions = new SuggestedActions
            {
                Actions = new List<CardAction>()
                {
                    new CardAction(type: ActionTypes.ImBack, title: MainStrings.HELP_BTN_TEXT_1, value: MainStrings.HELP_BTN_VALUE_1),
                    new CardAction(type: ActionTypes.ImBack, title: MainStrings.HELP_BTN_TEXT_2, value: MainStrings.HELP_BTN_VALUE_2),
                    new CardAction(type: ActionTypes.OpenUrl, title: MainStrings.HELP_BTN_TEXT_3, value: MainStrings.HELP_BTN_VALUE_3),
                },
            };

            return response;
        }

        public static IMessageActivity BuildHelpCard(ITurnContext turnContext, dynamic data)
        {
            var attachment = new HeroCard()
            {
                Title = MainStrings.HELP_TITLE,
                Text = MainStrings.HELP_TEXT,
            }.ToAttachment();

            var response = MessageFactory.Attachment(attachment, ssml: MainStrings.HELP_TEXT, inputHint: InputHints.AcceptingInput);

            response.SuggestedActions = new SuggestedActions
            {
                Actions = new List<CardAction>()
                {
                    new CardAction(type: ActionTypes.ImBack, title: MainStrings.HELP_BTN_TEXT_1, value: MainStrings.HELP_BTN_VALUE_1),
                    new CardAction(type: ActionTypes.ImBack, title: MainStrings.HELP_BTN_TEXT_2, value: MainStrings.HELP_BTN_VALUE_2),
                    new CardAction(type: ActionTypes.OpenUrl, title: MainStrings.HELP_BTN_TEXT_3, value: MainStrings.HELP_BTN_VALUE_3),
                },
            };

            return response;
        }

        public class ResponseIds
        {
            // Constants
            public const string Cancelled = "cancelled";
            public const string Completed = "completed";
            public const string Confused = "confused";
            public const string Greeting = "greeting";
            public const string Help = "help";
            public const string Intro = "intro";
        }
    }
}
